### PR TITLE
feat: implemented getters / setters and various string fns on Date 

### PR DESCRIPTION
**Pull request by ColinEberhardt**
This PR adds
 -  getters / setters for UTC dates (I don't plan to add any local-time related functionality)
 - `fromString`, as a substitute for the `new Date("2345-11-04")` style constructor - I resisted the temptation to use [as-regex](https://github.com/ColinEberhardt/assemblyscript-regex) for this task 😆 
 - `toISOString`

I've also modified `Date.UTC` so that it no longer calls out to the JS host.

### CODE REVIEW

File: std/assembly/date.ts
Comment (MaxGraey) [inline std/assembly/date.ts:239]: Could you split this to three separate functions:
`yearFromEpochDays`, `monthFromEpochDays` and `dayFromEpochDays`? This will make returning YMD object unnecessary

File: std/assembly/date.ts
Comment (MaxGraey) [inline std/assembly/date.ts:221]: What about reuse `daysInMonth` here as well?
```ts
export function daysInMonth(year: i32, month: i32): i32 {
  return month == 2
    ? 28 + i32(leapYear(year))
    : 30 + ((month + i32(month >= 8)) & 1);
}
```

File: std/assembly/date.ts
Comment (ColinEberhardt) [reply std/assembly/date.ts:239]: Most of the time you need multiple components (e.g. year and month), so splitting into separate functions would result in repeated calculations.

Is it better to provide use an array to return these values perhaps?

File: std/assembly/date.ts
Comment (dcodeIO) [inline std/assembly/date.ts:30]: There may be an opportunity here to make `fromString` basically allocation-free, by walking the string and using `strtol` to extract the values (currently doesn't support a limit, though). @MaxGraey Thoughts?

File: std/assembly/date.ts
Comment (MaxGraey) [reply std/assembly/date.ts:239]: I guess better manually inline everything. At leat for now until we have scalar replacement optimization.
@dcodeIO  wdyt?

File: std/assembly/date.ts
Comment (MaxGraey) [reply std/assembly/date.ts:30]: yeah, instead `split` better use `while + indexOf` loop here

File: std/assembly/date.ts
Comment (dcodeIO) [reply std/assembly/date.ts:239]: Another option could be to return just the `year` component, and put `month` and `day` into globals. So one still gets all the values, but without an allocation.

File: std/assembly/date.ts
Comment (dcodeIO) [reply std/assembly/date.ts:30]: Note that there would still be other avoidable allocations then when `substring`ing the parts of interest into `parseInt`. Hence I thought this may be a case for slightly extending `strtol` so it stops at a given limit :)

**Comment by bnbarak** [comment]
A tiny recommendations - add peace of mind for checking the odd corner cases in dates in different timezones. Perhaps add `moment.js` in devDependencies and add a test to compare the results of `assemblyscript Date() type` with `moment.js` Assuming `moment.js` is a source of truth to dates.

File: std/assembly/date.ts
Comment (ColinEberhardt) [reply std/assembly/date.ts]: done

File: std/assembly/date.ts
Comment (ColinEberhardt) [reply std/assembly/date.ts]: Rather than special-case year, I've modified the function to be void, returning all three values as globals https://github.com/AssemblyScript/assemblyscript/pull/1768/commits/03e4d8b8de9df68e6be98cef802225f5741034af

File: std/assembly/date.ts
Comment (ColinEberhardt) [reply std/assembly/date.ts:30]: I'll leave this one to you @MaxGraey  ;-)

**Comment by ColinEberhardt** [comment]
> A tiny recommendations - add peace of mind for checking the odd corner cases in dates in different timezones.

Thanks @bnbarak - I don't plan to add any timezone related functionality to `Date`, because JS dates are horrible! Instead I am implementing the Temporal spec here:

https://github.com/ColinEberhardt/assemblyscript-temporal

The only reason I am adding UTC-related `Date` functionality here is that because I need that for Temporal ;-)

File: std/assembly/date.ts
Comment (MaxGraey) [reply std/assembly/date.ts]: @ColinEberhardt ok

File: std/assembly/date.ts
Comment (MaxGraey) [reply std/assembly/date.ts:30]: @ColinEberhardt ok

**Review by dcodeIO — APPROVED** [review]
LGTM with the merge conflict fixed. We may also want to update [the docs](https://github.com/AssemblyScript/website/blob/master/src/stdlib/date.md) accordingly as a follow-up, and do a pass on `String#split` afterwards :)

**Comment by ColinEberhardt** [comment]
> LGTM with the merge conflict fixed

Thanks @dcodeIO  -  rebased onto `master` - so should be good to merge when the tests pass 🤞 

I'll update the docs after this is merged.

**Pull request merged**
